src/server/replica.cc

```diff
@@ -829,11 +830,22 @@ void DflyShardReplica::StableSyncDflyReadFb(Context* cntx) {
     if (!tx_data)
       break;
 
-    last_io_time_ = Proactor()->GetMonotonicTimeNs();
+    const bool is_ping = tx_data->opcode == journal::Op::PING;
+    if ((is_ping && tx_data->lsn == 0) || (!is_ping)) {
+      lsn_ += tx_data->journal_rec_count;
+    }
 
-    if (tx_data->opcode == journal::Op::PING) {
+    last_io_time_ = Proactor()->GetMonotonicTimeNs();
+    if (is_ping) {
       force_ping_ = true;
-      journal_rec_executed_.fetch_add(1, std::memory_order_relaxed);
+      if (tx_data->lsn != 0) {
+        const uint64_t expect = lsn_;
+        const bool is_expected = tx_data->lsn == expect;
+        LOG_FIRST_N(INFO, 10) << "tx_data->lsn=" << tx_data->lsn << " lsn_=" << expect;
+        DCHECK(is_expected) << "tx_data->lsn=" << tx_data->lsn << "lsn=" << expect;
+      } else {
+        journal_rec_executed_.fetch_add(1);
```

Comment:
This flow is so confusing. In replica takeover we send Op::PING but you dont fill the entry.lsn so it is 0 and therefor you end up increasing journal_rec_executed_

Reply:
yes, it's not ideal. If it's `0` it means that it's a Ping coming from `REPLTAKEOVER`. `LSN` start at 1, so `0` is used here to denote that the `PING` is used without the extension, that is `PING 0` == old `PING` and `PING >=1` == `PING LSN`. 

and also I treat `PING LSN`as separate and that's why they don't participate in incrementing journal_rec_executed.